clicommand/agent_start.go

```diff
@@ -1110,3 +1116,67 @@ func defaultSocketsPath() string {
 
 	return filepath.Join(home, ".buildkite-agent", "sockets")
 }
+
+// runAgentAPI runs an API socket that can be used to interact with this
+// (top-level) agent. It returns a shutdown function.
+func runAgentAPI(ctx context.Context, l logger.Logger, socketsPath string) func() {
+	path := agentapi.DefaultServerPath(socketsPath)
+	// There should be only one Agent API socket per agent process.
+	// If a previous agent crashed and left behind a socket, we can
+	// remove it.
+	os.Remove(path)
+	svr, err := agentapi.NewServer(path, l)
+	if err != nil {
+		l.Fatal("Couldn't serve Agent API socket: %v", err)
+	}
+
+	// Try to be the leader - no worries if this fails.
+	leaderPath := agentapi.LeaderPath(socketsPath)
+	if err := os.Symlink(path, leaderPath); err == nil {
+		l.Info("Agent API: This agent became leader")
+	}
+
+	// Whoever the leader is, ping them every so often as a health-check.
+	go leaderPinger(ctx, l, path, leaderPath)
+
+	return func() {
+		svr.Shutdown(ctx)
+		if d, err := os.Readlink(leaderPath); err == nil && d == path {
+			os.Remove(leaderPath)
+		}
+	}
+}
+
+// leaderPinger pings the leader socket for liveness, and takes over if it
+// fails.
+func leaderPinger(ctx context.Context, l logger.Logger, path, leaderPath string) {
+	pingLeader := func() error {
+		d, err := os.Readlink(leaderPath)
+		if err != nil {
+			// Not a symlink?
+			return err
+		}
+		if d == path {
+			// It's me! Don't bother pinging.
+			return nil
+		}
+
+		ctx, canc := context.WithTimeout(ctx, 100*time.Millisecond)
+		defer canc()
+
+		cli, err := agentapi.NewClient(ctx, leaderPath)
+		if err != nil {
+			return err
+		}
+		return cli.Ping(ctx)
+	}
+
+	for range time.Tick(100 * time.Millisecond) {
```

Comment:
TIL the assign-less `for range` expression! totally makes sense that you can do it though